learn/models/bootstrap-metrics/index.qmd

```diff
@@ -0,0 +1,212 @@
+
+For our application, we'll take the validation set predictions for each candidate model in the grid, bootstrap them, and then compute confidence intervals using the percentile method. 
+
+There's a tidymodels function called `int_pctl()` for this purpose. It has a method to work with objects produced by the tune package, such as our `mars_res` object. Let's compute 90% confidence intervals using 2,000 bootstrap samples:
+
+```{r}
+#| label: bootstrapping
+#| cache: true
+
+set.seed(140)
+mars_boot <- int_pctl(mars_res, alpha = 0.10)
+mars_boot
+```
+
+The results have columns for the mean of the sampling distribution (`.estimate`) and the upper and lower confidence bounds (`.upper` and `.lower`, respectively). Let's visualize these results: 
+
+```{r}
+#| label: int-plot
+
+mars_boot %>% 
+  ggplot(aes(num_terms)) + 
+    geom_line(aes(y = .estimate)) + 
+    geom_point(aes(y = .estimate), cex = 1 / 2) + 
+    geom_ribbon(aes(ymin = .lower, ymax = .upper), alpha = 1 / 5, fill = "blue") +
+    facet_wrap(~ .metric, scales = "free_y", ncol = 1) +
+    labs(y = NULL, x = "# Retained Terms")
+```
+
+Those are very tight! Maybe there is some credence to using many terms. Let's say that 40 terms seems like a reasonable value for that tuning parameter. It's large, but the out-of-sample statistics indicate that it does not overfit at this point. 
+
+## Test Set Intervals
+
+Suppose the MARS model was the best we could do for these data. We would then fit the model (with 40 terms) on the training set then finally evaluate the test set. tidymodels has an API called `last_fit()` that uses our original data splitting object (`delivery_split`) and the model specification. To get the test set predictions, we can use `collect_metrics()`: 
+
+```{r}
+#| label: test-res
+
+mars_final_spec <- 
+    mars(num_terms = 40, prod_degree = 2, prune_method = "none") %>% 
+  set_mode("regression")
+
+mars_test_res <- 
+  mars_final_spec %>% 
+  last_fit(time_to_delivery ~ ., split = delivery_split)
+
+collect_metrics(mars_test_res)
+```
+
+These values are pretty consistent with what the validation set (and its confidence intervals) produced: 
+
+```{r}
+#| label: test-compare
+mars_boot %>% filter(num_terms == 40)
+```
+
+`int_pctl()` also works on objects produced by `last_fit()`: 
+
+```{r}
+#| label: test-boot
+set.seed(168)
+mars_test_boot <- int_pctl(mars_test_res, alpha = 0.10)
+mars_test_boot
+```
+
+## Notes
+
+ - The point estimates produced by `collect_metrics()` and `int_pctl()` may disagree. They are two different statistical estimates of the same quantity. For reasonably sized data sets, they should be pretty close. 
+ 
+ - Parallel processing can be used for bootstrapping (using the [same tools](https://www.tmwr.org/grid-search#parallel-processing) that can be used with the tune package). 
```

Comment:
The tune page has some practical advice on setting those tools up: https://tune.tidymodels.org/articles/extras/optimizations.html#parallel-processing